packages/request-manager/src/controller.ts

```diff
@@ -136,96 +149,42 @@ export class TorController extends EventEmitter {
-        const bootstrap: BootstrapEvent[] = bootstrapParser(message);
-        bootstrap.forEach(event => {
-            if (event.type !== 'progress') return;
-            if (event.progress && !this.getIsBootstrapping()) {
-                // We consider that bootstrap has started when we receive any bootstrap event and
-                // Tor is not bootstrapping yet.
-                // If we do not receive any bootstrapping event, we can consider there is something going wrong and
-                // an error will be thrown when `maxTriesWaiting` is reached in `waitUntilAlive`.
-                this.startBootstrap();
-            }
-            if (event.progress === BOOTSTRAP_EVENT_PROGRESS.Done) {
-                this.successfullyBootstrapped();
-            }
-            this.emit('bootstrap/event', event);
-        });
-    }
-
-    public waitUntilAlive(): Promise<void> {
-        const errorMessages: string[] = [];
+    public async waitUntilAlive(): Promise<void> {
         this.status = TOR_CONTROLLER_STATUS.Bootstrapping;
-        const waitUntilResponse = async (triesCount: number): Promise<void> => {
-            if (this.getIsStopped()) {
-                // If TOR is starting and we want to cancel it.
-                return;
+
+        const abortController = new AbortController();
+        const checkConnection: ScheduledAction<boolean> = async signal => {
+            if (signal?.aborted) {
+                throw new Error('Tor controller check alive aborted');
             }
-            if (triesCount >= this.maxTriesWaiting) {
-                throw new Error(
-                    `Timeout waiting for TOR control port: \n${errorMessages.join('\n')}`,
-                );
+            const isConnected = await this.controlPort.connect();
+            const isAlive = this.controlPort.ping();
+            const isCircuitEstablished = this.getIsCircuitEstablished();
+            // It is running so let's not wait anymore.
+            if (isConnected && isAlive && isCircuitEstablished) {
+                return true;
             }
-            try {
-                const isConnected = await this.controlPort.connect();
-                const isAlive = this.controlPort.ping();
-                if (isConnected && isAlive && this.getIsCircuitEstablished()) {
-                    // It is running so let's not wait anymore.
-                    return;
-                }
-            } catch (error) {
-                // Some error here is expected when waiting but
-                // we do not want to throw until maxTriesWaiting is reach.
-                // Instead we want to log it to know what causes the error.
-                if (error && error.message) {
-                    console.warn('request-manager:', error.message);
-                    errorMessages.push(error.message);
+            throw new Error('Tor not alive');
+        };
+        const params: ScheduleActionParams = {
```

Comment:
This is much better, in my opinion. Just two things:

- `timeout` is a limit for one action attempt after which it throws instead, while `gap` is a delay between two consecutive attempts. It's perfectly fine to have both defined, I just want to make sure that it's what you want.
- I think there's no need for `abortController` check inside an action and aborting it from `attemptFailureHandler`, as the `getIsStopped` check could be done in the action itself, which makes `attemptFailureHandler` redundant.. If you're not against, I can try to propose an adjustment after the review.

Reply:
Would this work for you?
```typescript
        const checkConnection: ScheduledAction<boolean> = async () => {
            if (this.getIsStopped()) return false;
            const isConnected = await this.controlPort.connect();
            const isAlive = this.controlPort.ping();
            const isCircuitEstablished = this.getIsCircuitEstablished();
            // It is running so let's not wait anymore.
            if (isConnected && isAlive && isCircuitEstablished) {
                return true;
            }
            throw new Error('Tor not alive');
        };
        const params: ScheduleActionParams = {
            attempts: MAX_TRIES_WAITING,
            timeout: WAITING_TIME,
            gap: WAITING_TIME,
        };
```

Reply:
I understood the difference between `timeout` and `gap`, and think in this case both are necessary.

Regarding `abortController` if you want to propose and adjustment, it is welcome :)